Slit support

rupak0577 opened the pull request:
Adds Slit support to Cuts

rupak0577 commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -319,23 +330,54 @@ def stop(self):
+        if self.cut_axis_enabled:
+            path = image.get('path', None)
+            hdus = pyfits.open(path, 'readonly')
+            phd = hdus[0].data
```
Is there any other way of accessing the data?

ejeschke replied:
Can we assume the data is always in HDU 0?  If we are going to access the HDU directly in this plugin then maybe we need a GUI control for that.

I'm wondering if instead this data access should be abstracted behind `AstroImage` (or maybe a subclass) which keeps a reference to the full pyfits object.  Then `get_pixels_on_line()` could take an `hdu` and `axis` keyword parameters.

rupak0577 replied:
Hmm, I didn't think about that. So should it be something like MultiDim's HDU list? But that would mean opening the file directly which can then be used to access the data.

Cadair replied:
@ejeschke It seems that at the moment the `AstroImage` object doesn't hold a pointer to the whole array in the file, just the current 'image' slice. If there was a reference to the whole array that could be used for this it would be a lot easier.

In the absence of that, I think we need to open the correct HDU using the ginga file tools.

pllim replied:
This is a bit hacky, but I use it to open a given HDU automatically from within a plugin. This is a code snippet from a method that belongs to the plugin:
```python
class MyLocalPlugin(...):
    def redo(self):
        # stuff

        #----- Extract ERR extension -----
        # errsrc is ERR version of image
        imfile = image.metadata['path']
        imname = image.metadata['name'].split('[')[0]
        err_extnum = 2  # I have my own function to find it, but let's say it is always 2 here for simplicity
        if err_extnum < 0:
            errsrc = None
        else:
            chname = self.fv.get_channelName(self.fitsimage)
            chinfo = self.fv.get_channelInfo(chname)
            errname = '{0}[{1}]'.format(imname, err_extnum)
            if errname in chinfo.datasrc:  # ERR already loaded
                errsrc = chinfo.datasrc[errname]
            else:  # Force load ERR data
                errsrc = self.fv.load_image(imfile, idx=err_extnum)
                chinfo.datasrc[errname] = errsrc
                self.fv.make_callback('add-image', chinfo.name, errsrc)

        # more stuff
```

Cadair replied:
What does `self.fv.load_image` return?

pllim replied:
`AstroImage`

Cadair replied:
@pllim That won't work here because that only holds a 2D array, we need the full 3D or 4D array that is in the file.

pllim replied:
Is it possible to treat 3D as multiple 2D?

Cadair replied:
possible, yes, but not practical, as we need to slice the numpy array to efficiently access the slit data.

pllim replied:
Ah, okay.

ejeschke commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -319,23 +330,54 @@ def stop(self):
+                coords = np.array(coords)
+                slit = phd[:, coords[:, 1], coords[:, 0]]
+
+                plt.imshow(slit, cmap=plt.get_cmap('gray'), interpolation='nearest', origin='lower')
```
Is this supposed to pop up a new plot or modify the existing plot?  If the latter, the matplotlib OO interface should be used on self.plot.ax ...

rupak0577 replied:
Unfortunately, self.plot.ax.imshow(...) does not update the plot. I was wondering if the plot area could be split into two but that would scale down the plot.

ejeschke replied:
I think you need do something like `self.plot.fig.canvas.draw()` to refresh the plot after the `self.plot.ax.imshow()`.  You can do a `self.plot.ax.cla()` (it's something like that) to clear the axes prior to the imshow if necessary.

rupak0577 replied:
That works, thanks! But it acts weirdly and squeezes the Cuts plot.

Cadair replied:
if you are updating an existing plot you can do `im.set_array` (or `im.set_data` I can't remember). Which will not re-draw the whole figure, and therefore be faster. (where `im` is the `AxesImage` object returned from `imshow()`.

ejeschke commented:
@rupak0577, @Cadair, @pllim 
Now that Cuts responds to selecting slices in MultiDim (see commit 759a84d77730fd9c0ef991ad1417f80ce64a380a), do we need the cut axis bits here?  I'm thinking that the slice control should be orthogonal and not really baked into this plugin.

We could fairly easily relocate `MultiDim` to the left side panel--I'm thinking this is beneficial for other plugins that also `redo()` when the HDU or slice changes.

ejeschke commented:
I think ultimately we may want to look at simply making the image object hold the entire array (as suggested earlier in this thread) and select a slice via a `set_slice()` method or some such.  I'm worried a bit about large ramp frames, etc., but maybe there can simply be an option or preference.  Then plugins that want to change the slice can simply call `set_slice()` or similar on the image.  Might simplify `MultiDim` as well.

Cadair commented:
@ejeschke I think having the image object hold the whole array is the best way to go from a perspective of multi dimensional data. Generally you are going to want to be able to do operations on the whole array.

ejeschke commented:
@rupak0577, @Cadair 
Could you give commit 85c3c767689ce166328392da7e93b8372dbd9a5c a whirl?  This opens the MultiDim plugin in the lower left workspace, which means you can navigate through the slices quite easily  and watch your cut change, you can even watch the cut change using the animation controls (although I would recommend increasing the interval slightly to 0.4 or so, matplotlib plotting seems to be the bottleneck).

I'm wondering how much of the NAXIS navigation you need from this PR if that is available?

pllim commented:
@Cadair , I need clarification of "holding the entire array". I work with a format like the one below and I certainly do not want `ERR` and `DQ` loaded all the time.

EXT | EXTNAME |  Description
------|---------------|-----------------
0      | PRIMARY  |  Header
1      | SCI            |  2D float array
2      | ERR          |  2D float array
3      | DQ            |  2D int array

Cadair commented:
@pllim I mean the whole array for the current HDU. i.e. if EXT 0 is a 4D array there should be a reference available to the full array. SO you can slice, or find maxima or averages etc.

pllim commented:
@Cadair , thanks, that makes sense.

ejeschke commented:
@nabobalis, @Cadair, can you try the latest master (see comment above)?  Try using `MultiDim` and `Cuts` together--does it do more or less what you want?  I'm open to making MultiDim open by default when NAXIS>=3.

Regarding maintaining a handle to the full cube in the image--I'm going to look at that in a separate issue.

rupak0577 commented:
@ejeschke Cuts now responds to MultiDim so that fixes the issue.

nabobalis commented:
@ejeschke  Cuts now responds to multidim. 

But we want to be able to handle the full cube. We love space-time diagrams and the like in solar physics too much. This is what hopefully, this PR can address in some form.

ejeschke commented:
@Cadair, @nabobalis, @rupak0577 
Commit 7b8c22c9a7bc7b85d450a66bdb5a804a6711685c adds support for maintaining a handle to a multidimensional array.  See the changes to the [`set_naxis()`](https://github.com/ejeschke/ginga/blob/master/ginga/misc/plugins/MultiDim.py#L357) method in `MultiDim` for how to change the slice.  The full array can be obtained with the call to `get_mddata()` on the `AstroImage` object.

ejeschke commented:
Suggest this PR be reworked to reflect the changes in `MultiDim` and `Cuts`.  In particular, it strikes me that it might be better to subclass `Cuts` to add the slit view image and to better let you customize this for solar.  BTW, this is not uncommon.  We subclass a number of the plugins here and @pllim has done the same.

nabobalis commented:
So I take it that this isn't a common thing in astrophysics?

nabobalis commented:
By which I mean slits for astronomy data.

ejeschke commented:
@nabobalis, I'm not familiar with processing datacubes.  We have only one instrument that I know of that produces them regularly and I think there has been no demand for such a feature from them.  However, if this is useful more widely then I'm happy to have it added to `Cuts`.  Maybe @pllim can comment?

Is the slit image just a different way of slicing the cube than you get with the naxis controls in `MultiDim`?

nabobalis commented:
Just about, we want to slice the array such that we get a distance time graph. Basically, we take cut and extend it into time.

ejeschke commented:
Ok, let's just rework for the changes and see where we are.  Unless the new graphs/images are done as popups the `Cuts` gui might get a little crowded.  Another option is to use a tabbed view like in `Pick` or a scrolled view, etc.

I'm reviewing your Crispex screenshot.  Do you want to reproduce the top right-hand plot as well?

nabobalis commented:
Yes to that question. we have partial support for it. Isn't a spectrum view nice for astrophysics? I feel kind of bad adding solar specific features!

ejeschke commented:
@nabobalis, I suspect that you are right that it is very useful.  It's just that I'm not so familiar with this kind of data processing.  However, I'm game to add any features that are useful.  Just want to be able to collapse GUIs that are only sometimes used.  We use `Cuts` here mostly as a quick look for judging optical/IR spectrograph images.

ejeschke commented:
Non-cubes, in the usual case.

ejeschke commented:
Probably we could add a tabbed view for the current plot, which would allow the "spectral intensity vs. position" graph to occupy the same space.  If you are ok with it then we could easily add the slit image as yet another tab.  This would push the cuts controls down to the bottom of the panel but allow the slit image to be tall.  Or we could make the slit image a pop out that you get by pressing a button.

nabobalis commented:
Yeah either a tab or a pop out window once you involve a slit would be nice. As long as its got enough space.

pllim commented:
I currently do not use Ginga with data cubes, so I cannot comment much on this. But development wise, it makes more sense to sub-class first and get it working for your particular dataset. And then if there is enough demand from general Ginga users, incorporate that into core Ginga distribution.

pllim commented:
Just as in `astropy`, they have the core distribution, and they have affiliated packages for specialized functionalities.

Cadair commented:
I know @astrofrog and @keflavich have been doing lots of work on astro radio cubes, I wonder if they have anything to say about the general use of viewing them in ginga.

keflavich commented:
@Cadair - what do you want to know?  Slicing across cubes (radio or optical - MUSE and KMOS produce them in optical/NIR, ALMA/EVLA/single dish in the radio) is a very common operation, but examining a single spectrum (or an average over some region) is crucial.  Swapping the displayed axes is very helpful and supported by few tools.

Cadair commented:
so we are suggesting adding an extension to the cuts functionality of ginga, which currently plots the intensity of the image along a drawn line, to being able to plot an image of the values of the cut along the third axis of the array. so for a x,y,t cube, it would be a distance-time diagram, or for a x,y,lambda cube it would be a distance-wavelength graph.

It sounds like this is something that would be useful for general astro cubes? 

When you say single spectrum you mean for a single pixel?

And when you say swapping the displayed axes do you mean making the main image say x,lambda? for a given y value?

keflavich commented:
> It sounds like this is something that would be useful for general astro cubes?

Yes!

> When you say single spectrum you mean for a single pixel?

Yes

> And when you say swapping the displayed axes do you mean making the main image say x,lambda? for a given y value?

And, yes.

Taking a slice at off-axis angles is also useful, but computationally expensive and therefore slow.  Presumably, the `slice` interface will come with some option for moving the slice (the long-slit) and then re-extracting?  It would be nice if that was configurable, either to update on-the-fly or on some keystroke.

Cadair commented:
So this PR doesn't do the axes swapping, although that is something that could probably be done quite easily with the recent changes to ginga. I will create a new issue for that.

This PR does do the other two things though, single spectrum for an arbitrary line, and all spectra for an arbitrary line (as an image). I think the current `Cuts` interface to ginga allows you to move the cut and then the spectra would update once at the end of the move operation (is this correct @rupak0577 @nabobalis ?).

nabobalis commented:
@cadair, I think so. But it had to be checked. Also it updating with the mouse needs to be added.

Cadair commented:
What updating with the mouse?

nabobalis commented:
The spectra at a pixel. Unless it click to update on pixel?
On 30 Jul 2015 3:56 pm, "Stuart Mumford" <notifications@github.com> wrote:

> What updating with the mouse?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ejeschke/ginga/pull/129#issuecomment-126359011>.
>

Cadair commented:
I thought we were putting that in the zoom plugin?

nabobalis commented:
I've lost track of where that is going.

Cadair commented:
well we can work it out, but we want to do mouse over somehow.

rupak0577 commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -221,6 +235,29 @@ def build_gui(self, container):
+            # Add callbacks
+            for chkbox in self.axes_chkboxes:
+                cbox = chkbox.get('widget')
+                cbox.add_callback('activated', lambda w, tf: self.axis_toggle_cb(w, tf, chkbox))
```
@ejeschke Somehow, this does not work - only the last checkbox works. Is there any other way to add the callbacks?

ejeschke replied:
@rupak0577, are you trying to add radio button like behavior with exclusivity?  If so, please use the `Widgets.RadioButton`.  For second and remaining buttons you specify the "group" using the group parameter which refers to the first widget in the group.  Also, you *may* get a callback with `False` when the widget has toggled "off" so you want to check for True before you actually do something.  You can see examples in the `Cuts` plugin radio button processing.

rupak0577 replied:
@ejeschke I am trying to add check boxes according to the number of axes so that multiple axes can be used for slicing. What I can't figure out is how to add a callback for each checkbox.

ejeschke replied:
If they are independent checkboxes you certainly should be able to assign separate callbacks for them.  If they don't fire when toggled then there may be a bug.  Do you get the same behavior with gtk?

rupak0577 replied:
@ejeschke The `state` of the checkboxes does not change except the last one. I think that is due to the loop that adds the callback to each checkbox. There is no problem with the GUI.

ejeschke commented:
@rupak0577, this PR causes the ordinary Cuts on a 2D image to fail, because it now attempts to access a 3D image in `_plotpoints()` (for the slit).

ejeschke commented:
Similarly, I tried Cuts on a 4D image and that also fails in the same spot.  I think the code handling slit needs to check explicitly for a 3D object before trying to slice it in 3D.

rupak0577 commented:
@ejeschke Fixed the slicing issue. 

I was wondering if the plot could be made bigger by replacing the instructions textbox with a collapsible widget.

ejeschke commented:
@rupak0577, if you know of a good collapsible widget I would like to hear of it.  I think the Qt QDockWidget has some capabilities like that, but it only works in certain places, and may not work in the general case for plugins.  So I think it may be necessary to construct a custom container widget.  In any case, I have long wanted something like that for the plugins.  We would need a gtk solution as well.

ejeschke commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -204,6 +217,15 @@ def build_gui(self, container):
+        elif len(image.naxispath) > 0:
+            self.mdfile = True
+            self.build_axes(vbox)
+
```
`redo()` is called when a new image is loaded into the viewer, so this test needs to be performed there as well, *before* the calls to replot.

ejeschke commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -227,6 +249,22 @@ def build_gui(self, container):
+        hbox = Widgets.HBox()
+        hbox.set_border_width(4)
+        hbox.set_spacing(1)
+
```
Since the axes checkboxes are only used for the slit, and only useful if there is a 3D cube, I recommend moving the checkbox array into the bottom of the slit tab instead of the main window.  I suggest building the hbox holder in `build_gui` and just keeping a handle to it, so that you can just clear it and repopulate it with new checkboxes when a new image arrives.

ejeschke commented:
Another possibility for making more room for the slit image is to make the tabs encompass more controls.  So, e.g. the Cuts tab could include all the cuts controls.  The `Close` button always needs to be visible, but this would give ample vertical room for the slit plot to expand inside the Slit tab.

rupak0577 commented:
@ejeschke Updated. 

GTK has the Expander Widget but Qt does need a custom one. I would like to work on it but probably after the coding period if that's okay.

Regarding clearing the axes in `redo()` - since `draw_cb()` also calls `redo()`, the checkboxes get cleared after a cut. So how do I check if a new image has been loaded?

ejeschke commented:
@rupak0577, this is looking better.  I still get an error in the `_plot_slit()` code when the cuts line is moved so that it falls off the edge of the data.  I think you need to basically throw out any coords that lie outside the data boundaries of the slit slice (this is what happens for plotting the regular cut--you can see that it still works if the line goes off the edge).

BTW, I think `_plot_slit()` should be called *not* from `_plotpoints()`, but as a follow-on call in `_redo()`.

Regarding `redo()`: that is the official method that is called when a "new image" enters the viewer or the image changes.  So I think maybe `draw_cb()` and `redo()` need to be restructured so that you can do the right kind of check for the data axes changing in `redo`.  I'll take a look at the master branch and maybe ask you rebase against a change to allow that.

ejeschke commented:
@rupak0577, commit fb27fcd3ad82364c22fbb926d4f1dc171a6522d3 refactored so that the replotting is separated from the `redo()` method.  This should allow you to rebuild the axes checkboxes when the data changes.  Can you rebase against that please.

ejeschke commented:
Also, I notice that you have changed the code so that you are plotting the cuts directly through the matplotlib axes instead of through the `Plot` object.  This is ok, but we've now lost the "tight_layout" call and so the axes labels are slightly occluded now.  It would be nice to fix that if possible--maybe by adding some padding into the matplotlib axes?

Cadair commented:
You can call `tight_layout ` on the `Figure` object if you have access to that.

ejeschke commented:
Yeah, figure is accessible.  `tight_layout` sometimes gives a little bit too much padding, as we've seen with the gtk version.

Cadair commented:
Fair enough, I think you can adjust the passing with arts to `tight_layout `

ejeschke commented:
@rupak0577, check out commit be1a006c2178f1f7da906ca9ee6908e1dfad3ad3.  This adds support for the Expander widget.  The Qt solution is kind of hacky, so if you want to improve it later, please do.  I've changed the Cuts plugin to use it as an example.

rupak0577 commented:
@ejeschke Thanks! It works perfectly.

I have reverted the plotting (don't know why it wasn't working before) .

Regarding the error in `_plot_slit()`, I am not getting an error although it's true that the plotting should not work for points outside the boundaries.

rupak0577 commented:
@ejeschke Added check for points outside image and support for bezier cut.

ejeschke commented on ginga/canvas/types/basic.py:
```diff
@@ -573,11 +573,15 @@ def select_contains(self, viewer, data_x, data_y):
+                    for x, y in self.get_points_on_curve(image) ]
+        else:
+            res = [ [x, y] if 0 <= x < wd and 0 <= y < ht else numpy.NaN
+                    for x, y in self.get_points_on_curve(image) ]
```
Hmm... not sure if you want `NaN`s mixed in with coordinate pairs.  This would cause basic numpy array transposition to fail.  Maybe we need a "magic" bad coordinate value.  At the very least this needs to return a pair instead of `NaN`.

rupak0577 replied:
The NaNs are removed before plotting in the bezier part of `_plot_slit()`

ejeschke commented:
@rupak0577, I think we are pretty close on this one. @nabobalis, @Cadair, can you guys have a look and give your feedback?

I'm thinking that maybe at least the Move/Draw/Edit buttons ought to fall below the tabs (not embedded in the first tab) so that even if you are mostly using the slit tab you can easily move between moving, drawing and editing the slit line.

rupak0577 commented:
Yeah, that would be nice. Probably the cut selection box too.

This still needs to be tested with a 4D file.

ejeschke commented:
If you are going to move the Cut selection box down then may as well move all the Cut controls.  If the typical use case is that a single slit line will be created and then redrawn/moved/edited it will not be necessary to access that so frequently.

Cadair commented:
@rupak0577 

I have just tested this with a dummy 4D file. A few comments are below.

It seems that the slit is not re-drawn when you change from NAXIS3 to NAXIS4. Also you should not be able to select both NAXIS3 and NAXIS4, if you start with NAXIS3 selected when you select NAXIS4 it should de-activate 3 automatically.

Also, the yaxis should not say "Time" unless we are going to start determining the coordinate system and type for the 3rd or 4th dimensions it should probably say nothing, unless someone has a better idea.

It would also be nice if you could transpose the plot with a checkbox, personally I would think it would be more logical to have the slit position up the y axis but that could just be me.

Cadair commented:
Also @rupak0577 The whole cuts control should be where the draw move and edit buttons now are, because you might want to make a new slit just like you make a new cut.

Cadair commented:
Also, it seems that slit does not work with the freepath mode, I just seem to be getting one pixel being repeated.

rupak0577 commented:
@Cadair Updated.

Cadair commented:
@rupak0577 Nice! Thanks.

One thing, you should keep the axis label that is "position along slit" so that you know what axes is the position and which is the dependant variable.

Otherwise I think I am :+1:  on this.

rupak0577 commented:
@Cadair Adding the axis label seems difficult since it transposes the plot by transposing the array. So the axis must be manually flipped. Also the "Slit position" axis is different for NAXIS3 and NAXIS4 (for 3 it's along x-axis, for 4 it's along y).

Cadair commented:
hmmm, ok.

The way I see to be able to do it is to work out which axis of the array is the same length as the number of slit coordinates and then decide on the label that way.

i.e.

```python
if len(coords) == slice.shape[0]:
    ax.set_ylabel('Position along slit.")
else:
    ax.set_xlabel("Position along slit.")
```

rupak0577 commented:
@Cadair Thanks! It's somewhat hacky but works (only for axis 3,4 though).

This looks complete now.

ejeschke commented:
@nabobalis, any comment?  Otherwise I think we could merge this.

nabobalis commented:
@ejeschke We had a few small bugs with it the last time we used it. I'll check to see if they are still there.

rupak0577 commented:
@nabobalis @ejeschke One remaining issue is making MultiDim work with Slit. The checkboxes are cleared and whenever the image changes. Finding out whether a new image has been loaded or a new slice seems difficult.

nabobalis commented:
@rupak0577 Using your latest branch for this. When you create a slit, a cut is also created. Thus when you save out the data, it saves out both sets of data. Not sure if there is a way to suppress the cut. 

With regards to the Multidim slider, I guess if the NAXIS button does not clear when you move the slider would be a quick solution to that issue. 

@ejeschke Are you aware of a way to extract a fits header for a cut? It would be nice to save out the slit data as a fits file.

ejeschke commented:
@rupak0577, can you rebase against master.  I think we are close to merging this.  I'll take a look at the checkboxes being cleared issue.

rupak0577 commented:
@ejeschke Updated.

The plugin seems a little crowded now. It would also be nice to have the plot remain constant in size. So I was wondering if it would be better to have the plugins as separate floating windows?

nabobalis commented:
Maybe have the option of having the new plots be popped out into a new window. But that would be in the preferences. 

I assume the issue of crowding is due to peoples screen size/resolutions.

rupak0577 commented:
@nabobalis Save and checkbox clearing issues fixed.

ejeschke commented:
@rupak0577, I think there are a couple of minor issues still.  

1) if I have slit enabled by choosing an axis in the Slit tab and then press "Play" in MultiDim to step through the cube rapidly at the default update rate (0.10) the displays do not seem to keep up and the viewer becomes unresponsive.  This doesn't happen in Cuts master so I think it is just due to the extra work being done each step to extract and update the slit image (many things are being updated at each step).  I think at the least we need to set the default rate to something lower.  On my oldish machine 0.3 or 0.4 rate it seems to keep up ok.   Not sure what the best way to handle this because performance of algorithms always depends on the users hardware to some degree, but I guess we don't want to provide a poor experience for average current hardware out of the box.

2) I think we need to figure out a way to keep the axis controls from being rebuilt on each step through the same cube.  After all they are not changing unless we change HDUs or load a new file.  I'll take a look at this problem.

rupak0577 commented:
@ejeschke Strangely, it happens with me for Cut master. It works fine if there is no cut but if a cut is present, it works for a while before the viewer hangs. So I guess the problem lies with the viewer. Anyways, I will try to look for a solution.

ejeschke commented:
@rupak0577, sorry, what happens with Cut master?  Is it possible to consistently reproduce the viewer hang?  If so, could you open an issue and include instructions on how to produce the hang?  Thanks!

rupak0577 commented:
@ejeschke Yeah, the viewer freezes but the cuts plot continues to update itself. Buttons are unresponsive as well.
It happens at the default rate 0.10 and the size is 928x548. Somewhat inconsistent though - it either hangs early or it takes a while.

ejeschke commented:
@rupak0577, what about if the rate is 0.4 or so?

rupak0577 commented:
@ejeschke Yeah, that works fine

rupak0577 commented:
@ejeschke @nabobalis Fixed the performance issues with MultiDim

nabobalis commented:
If I create a slit and move the NDim Slider, ginga freezes. It will crash for me with this
```
RuntimeError: maximum recursion depth exceeded while calling a Python object
```

My maximum recursion depth is set quite high as well.

ejeschke commented:
@rupak0577, I get the same behavior as @nabobalis.  If I don't create a slit (just a cut), and drag the slider (or play) things are ok.  If I enable a slit the viewer becomes very slow or non-responsive.

rupak0577 commented:
@ejeschke Added a few fixes - basically removed rebuilding of axes checkboxes and added a check so that it draws only if MultiDim is not playing. Should work smoothly now.

@nabobalis Sorry, I can't seem to reproduce your error - the slider and the buttons work perfectly for me with no crash or freeze. I have added a few more fixes so check out the latest update and see if it works.

nabobalis commented:
This now works without the issues I had previously. WIth regards to speed, without benchmarks, its hard to prove. 

@rupak0577 Anything else you wanted to do this? Or you happy for it to be merged?

rupak0577 commented:
@nabobalis @ejeschke If nothing else is required, we can merge it

nabobalis commented:
@ejeschke Are you happy with this currently? I think it is ready to be accepted.

ejeschke commented:
I'm getting the following traceback in the log:
2015-10-15 16:48:18,444 | E | Callback.py:88 (make_callback) | Error making callback 'draw-event': 'Cuts' object has no attribute 'mdim_ref'
2015-10-15 16:48:18,444 | E | Callback.py:89 (make_callback) | Traceback:
  File "/home/eric/Git/Ginga/ginga/misc/Callback.py", line 72, in make_callback
    res = method(*cb_args, **cb_kwdargs)

  File "/home/eric/Git/Ginga/ginga/misc/plugins/Cuts.py", line 908, in draw_cb
    return self.replot_all()

  File "/home/eric/Git/Ginga/ginga/misc/plugins/Cuts.py", line 682, in replot_all
    if self.cutstag != self._new_cut and not self.mdim_ref.is_playing():

ejeschke commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -399,8 +483,22 @@ def redo(self):
+        chname = self.fv.get_channelName(self.fitsimage)
+        chinfo = self.fv.get_channelInfo(chname)
+        self.mdim_ref = chinfo.opmon.getPlugin('MultiDim')
+
```
I don't think `mdim_ref` is getting set if you open `Cuts` *after* loading an image.  Then it raises an exception (I've marked it lower down) because it is not defined.

In general, I think the approach of trying to check what another plugin is doing and work around that is probably not the correct approach.

ejeschke commented on ginga/misc/plugins/Cuts.py:
```diff
@@ -514,10 +677,10 @@ def replot_all(self):
+
+        # Redraw slit image for selected cut
+        if self.cutstag != self._new_cut and not self.mdim_ref.is_playing():
+            self.redraw_slit('redraw')
```
Here `mdim_ref` is not defined in the case I described above (open Cuts plugin after opening the file), so I get an exception raised and the Slit is not redrawn.

ejeschke commented:
I'm thinking that for playback, it may be better to support this via a fix to the viewer *not* to make the `image-set` callback which triggers all of the plugins, especially ones that plot things with matplotlib, because those tend to slow everything down.  Then playback would support a pretty high rate (I have a test program that basically can watch a HD video in a ginga widget at 30FPS).  You could still use the buttons to step through with the callback applied.

rupak0577 commented:
@ejeschke  Would that be a fix to `ImageView`? I think `imshow` is the bottleneck and I have been trying to use `set_data` instead but it does not seem to work.

rupak0577 commented:
@ejeschke @nabobalis Removed Multidim check. Ready to be merged(ignoring speed issues).

ejeschke commented:
@rupak0577, how about using a Ginga widget for the slit image?  It will certainly be faster than matplotlib.  Take a look at `Pick` if you need an example of how to create a small ginga widget in a plugin.

ejeschke commented:
Let me know if you want help.

rupak0577 commented:
@ejeschke How can it be used to set the image data(instead of `imshow`)? Also, how can I add the axis labels?

ejeschke commented:
Hmm, I see your point.  Plus we'd have to make sure the "save" bits are all supported.  Perhaps a future upgrade of the plugin if we get some basic plot axes for ginga.

rupak0577 commented:
@ejeschke It isn't the perfect solution but it would work if we plot depending upon the tab the user is in - if the current tab selected is Cuts then plot otherwise don't. But the best way is to change the data instead of calling `imshow` each time.

ejeschke commented:
I've fixed up some problems with this PR in commit 376bf76ebeabb3235b41bb0d83602b11f26fe573.  I'm still seeing plenty of problems interoperating with 2D images, so I've made the Slit functionality a preference (default disabled) for now.  Just add `enable_slit = True` to your `$HOME/.ginga/plugin_Cuts.cfg` file.  I'll start an issue with the problems I'm seeing.

Cadair commented:
@ejeschke did you open those issues?

ejeschke commented:
Sorry, not yet.  I'll make a note to do it.  You can see what kinds of errors you get by enabling the preference and just loading a 2D image.  Ideally, the slit features should just silently bypass if they don't work for 2D, not generate error messages or worse, cause the Cuts plot to fail.
